File: Arrays & Strings/#3606 - Coupon Code Validator - Easy/Solution.py
class Solution:
    def validCoupons(self, code, businessLine, isActive):
        valid_categories = {"electronics", "grocery", "pharmacy", "restaurant"}
        category_order = {
            "electronics": 0,
            "grocery": 1,
            "pharmacy": 2,
            "restaurant": 3
        }

        valid_list = []
        for c, b, active in zip(code, businessLine, isActive):
            if (active and
                b in valid_categories and
                len(c) > 0 and
                all(ch.isalnum() or ch == '_' for ch in c)):
                valid_list.append((b, c))

        valid_list.sort(key=lambda x: (category_order[x[0]], x[1]))
        return [c for _, c in valid_list]

File: Arrays & Strings/#3606 - Coupon Code Validator - Easy/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<string> validCoupons(vector<string>& code, vector<string>& businessLine, vector<bool>& isActive) {
        unordered_set<string> validCategories = {"electronics", "grocery", "pharmacy", "restaurant"};
        unordered_map<string, int> categoryOrder = {
            {"electronics", 0},
            {"grocery", 1},
            {"pharmacy", 2},
            {"restaurant", 3}
        };

        vector<pair<string, string>> validList;
        for (int i = 0; i < code.size(); i++) {
            bool validCode = !code[i].empty() && all_of(code[i].begin(), code[i].end(),
                                [](char ch) { return isalnum(ch) || ch == '_'; });

            if (isActive[i] && validCategories.count(businessLine[i]) && validCode) {
                validList.push_back({businessLine[i], code[i]});
            }
        }

        sort(validList.begin(), validList.end(), [&](auto &a, auto &b) {
            if (categoryOrder[a.first] != categoryOrder[b.first])
                return categoryOrder[a.first] < categoryOrder[b.first];
            return a.second < b.second;
        });

        vector<string> result;
        for (auto &p : validList) result.push_back(p.second);
        return result;
    }
};
